### PR TITLE
#1842 Add Currying Design Pattern

### LINKED ISSUE #1842: Currying
I's a functional programming design pattern. As far as I understand you call a function with less parameters than usual and it can be used to defer processing when it has many steps involved.

UML example

```
@startuml
'https://plantuml.com/class-diagram

interface BiFunction <ParamType1, ParamType2, ResultType>{
}

interface Curried<ParamType1, ParamType2 extends Simple> {
  +ParamType2 apply(final ParamType1 a);
}

interface Simple<ParamType1, ParamType2, ResultType> extends BiFunction {
  +ResultType apply(final ParamType1 a, final ParamType2 b);
}

@enduml
```

image

With the curried version you can do stuff like:

```
   System.out.println(curriedAdd.apply(4).apply(5));

    Function1<Integer, Integer> adder5 = curriedAdd.apply(5);
    //some complicate and time consuming process here
    System.out.println(adder5.apply(4));
    System.out.println(adder5.apply(6));

```

### PR DESCRIPTION
**Pull request by hkattt**
This PR resolves #1842.

I adapted the currying example suggested in #1842. My example focuses on using a curried builder function to create `Book` objects. I added unit tests for the `App` and `Book` classes. I filled in the `README` template as specified [here](https://github.com/iluwatar/java-design-patterns/wiki/02.-Pattern-template). 

Things to look at:
- What is the appropriate [tag](https://github.com/iluwatar/java-design-patterns/wiki/07.-Categories-and-Tags) for the currying design pattern?
- I left out the 'related patterns' section from the `README` since I could not find any related patterns. Is this okay?
- Have I listed a sufficient amount of known uses?

### CODE REVIEW

File: currying/README.md [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+## Consequences
+Pros
+* Currying allows for partial application, which can be used to create specialised functions concisely.
+
+Cons
+* The order of the parameters in a curried function is important since we want to take advantage of partial application. It is best to input the most general parameters first and input specific parameters last. 
+* As shown in the programmatic example above, curried functions with several parameters have a cumbersome type signature (in Java).
+
+## Credits
+* [Currying in Java](https://www.baeldung.com/java-currying)
```
Comment (Jesonzavic) [inline currying/README.md]: For related patterns, perhaps the builder pattern and functional interfaces patterns you mentioned within your write up?

File: currying/README.md [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+
+public interface AddTitle {
+Book.AddPublicationDate withTitle(String title);
+}
+
+public interface AddPublicationDate {
+Book withPublicationDate(LocalDate publicationDate);
+}
+```
+The semantics of the `builder` function can easily be understood. The `builder` function returns a function `AddGenre`, which adds the genre to the book. Similarity, the `AddGenre` function returns another function `AddTitle`, which adds the title to the book and so on, until the `AddPublicationDate` function returns a `Book`. The below example demonstrates how partial application can be used with the `builder` function to create specialised book builder functions.
```
Comment (Jesonzavic) [inline currying/README.md]: I feel a little bit lost here, perhaps add a very short example with the text to create a single book?

File: currying/pom.xml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        <artifactId>java-design-patterns</artifactId>
+        <version>1.26.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>currying</artifactId>
+    <dependencies>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
```
Comment (Jesonzavic) [inline currying/pom.xml]: This should not depend on versioning, and if i recall correctly, you should be able to use junit.jupiter for most of the things you would need junit for.

**Review by Jesonzavic — CHANGES_REQUESTED** [review]

File: currying/README.md (hunk 1 above)
Comment (hkattt) [reply currying/README.md]: This is a good point. I will add the builder pattern. However, I don't think functional interfaces are a pattern. So I will not include them.

**Comment by hkattt** [comment]
Hey, @Jesonzavic I added the changes that you asked for.

**Review by Jesonzavic — APPROVED** [review]

**Review by ANUEthan — APPROVED** [review]
Informative, thorough documentation and well implemented design pattern.
All methods are either accompanied by Javadoc comments or have self-evident names, well done!

**Pull request merged**